lightningd/bitcoind.c

```diff
@@ -784,6 +785,124 @@ void bitcoind_gettxout(struct bitcoind *bitcoind,
+			       const struct bitcoin_tx_output *output,
+			       void *arg)
+{
+	struct filteredblock_call *call = (struct filteredblock_call *)arg;
+	struct filteredblock_outpoint *o = call->outpoints[call->current_outpoint];
+
+	/* If this output is unspent, add it to the filteredblock result. */
+	if (output)
+		tal_arr_expand(&call->result->outpoints, tal_steal(call->result, o));
+
+	call->current_outpoint++;
+	if (call->current_outpoint < tal_count(call->outpoints)) {
+		o = call->outpoints[call->current_outpoint];
+		bitcoind_gettxout(bitcoind, &o->txid, o->outnum,
+				  process_getfilteredblock_step3, call);
+	} else {
+		/* If there were no more outpoints to check, we call the callback. */
+		call->cb(bitcoind, call->result, call->arg);
+		tal_free(call);
+	}
+}
+
+static void process_getfilteredblock_step2(struct bitcoind *bitcoind,
+					   struct bitcoin_block *block,
+					   struct filteredblock_call *call)
+{
+	struct filteredblock_outpoint *o;
+	struct bitcoin_tx *tx;
+	call->result->prev_hash = block->hdr.prev_hash;
+
+	/* Allocate an array containing all the potentially interesting
+	 * outpoints. We will later copy the ones we're interested in into the
+	 * call->result if they are unspent. */
+
+	call->outpoints = tal_arr(call, struct filteredblock_outpoint *, 0);
+	for (size_t i = 0; i < tal_count(block->tx); i++) {
+		tx = block->tx[i];
+		for (size_t j = 0; j < tx->wtx->num_outputs; j++) {
+			const u8 *script = bitcoin_tx_output_get_script(NULL, tx, j);
+			if (is_p2wsh(script, NULL)) {
+				/* This is an interesting output, remember it. */
+				o = tal(call->outpoints, struct filteredblock_outpoint);
+				bitcoin_txid(tx, &o->txid);
+				o->satoshis = bitcoin_tx_output_get_amount(tx, j);
+				o->txindex = i;
+				o->outnum = j;
+				o->scriptPubKey = tal_steal(o, script);
+				tal_arr_expand(&call->outpoints, o);
+			} else {
+				tal_free(script);
+			}
+		}
+	}
+
+	call->result->outpoints = tal_arr(call->result, struct filteredblock_outpoint *, 0);
+	call->current_outpoint = 0;
+	if (tal_count(call->outpoints) == 0) {
+		/* If there were no outpoints to check, we can short-circuit
+		 * and just call the callback. */
+		call->cb(bitcoind, call->result, call->arg);
+		tal_free(call);
+	} else {
+
+		/* Otherwise we start iterating through call->outpoints and
+		 * store the one's that are unspent in
+		 * call->result->outpoints. */
+		o = call->outpoints[call->current_outpoint];
+		bitcoind_gettxout(bitcoind, &o->txid, o->outnum,
+				  process_getfilteredblock_step3, call);
+	}
+}
+
+static void process_getfilteredblock_step1(struct bitcoind *bitcoind,
+					   const struct bitcoin_blkid *blkid,
+					   struct filteredblock_call *call)
+{
+	/* So we have the first piece of the puzzle, the block hash */
+	call->result->id = *blkid;
+
+	/* Now get the raw block to get all outpoints that were created in
+	 * this block. */
+	bitcoind_getrawblock(bitcoind, blkid, process_getfilteredblock_step2, call);
```

Comment:
How does this interact with pruning BTW?

Reply:
It doesn't change the requirement for the blocks actually being present in `bitcoind`. Before we were asking `bitcoin-cli getblock xyz true` which parses the block in `bitcoind` and returns just a list of transaction IDs, now we call `getblock xyz false` and parse in `lightningd`. If the block we were looking for was pruned away we'd fail the lookup and discard the announcement. Now we do the same. Hence the warning that pruned nodes are not supported :-)

The reason I built `getfilteredblock` as a pseudo `bitcoin-cli` command is that this should make it easier to replace in a plugin with a custom method of retrieving blocks (web service, cache, ...) :wink: